intelmq/lib/bot_debugger.py

```diff
@@ -0,0 +1,81 @@
+ * starts the bot as is (default)
+ * processes single message, either injected or from default pipeline (process subcommand)
+ * reads the message from input pipeline or send a message to output pipeline (message subcommand)
+"""
+import logging
+import json
+
+from intelmq.lib.message import Event
+from importlib import import_module
+from intelmq.lib.utils import StreamHandler
+from intelmq.lib.message import Event
+
+
+class BotDebugger:
+        def leverageLogger(self, level = logging.DEBUG):
+            self.instance.logger.setLevel(level)
+            for h in self.instance.logger.handlers:
+                if isinstance(h, StreamHandler):
+                    h.setLevel(level)
+
+        def __init__(self, module_path, bot_id, run_subcommand = None, message_kind = None, dryrun = None, msg = None):
+            module = import_module(module_path)
+            bot = getattr(module, 'BOT')
+            self.instance = bot(bot_id)
+            self.leverageLogger()
+
+            if not run_subcommand:
+                self.instance.start()
+            else:
+                self.instance._Bot__connect_pipelines()
+                if run_subcommand == "process":
+                    self._process(dryrun, msg)
+                elif run_subcommand == "message":
+                    self.leverageLogger(level = logging.INFO)
+                    self._message(message_kind, msg)
+                else:
+                    print("Subcommand {} not known.".format(run_subcommand))
+
+        def _process(self, dryrun, msg):
+            if msg:
+                self.instance._Bot__source_pipeline.receive = lambda: msg
+                self.instance.logger.info("Message from cli will be used when processing.")
+
+            if dryrun:
+                self.instance.send_message = lambda msg: self.instance.logger.info("DRYRUN: Message would be sent now!")
+                self.instance.acknowledge_message = lambda: self.instance.logger.info("DRYRUN: Message would be acknowledged now!")
+                print("Dryrun only, no message will be really sent through.")
+
+            self.instance.logger.info("Processing...")
+            self.instance.process() # normal run - bot loads and parses its next msg from pipeline
+
+        def _message(self, message_action_kind, msg):
+            if message_action_kind == "get":
+                self.instance.logger.info("Trying to get the message...")
+                msg = self.instance.receive_message()
+                print(msg)
+            elif message_action_kind == "pop":
+                self.instance.logger.info("Trying to pop the message...")
+                print(self.instance.receive_message())
+                self.instance.acknowledge_message()
+            elif message_action_kind == "send":
+                if msg:
+                    try:
+                        msg = Event(json.loads(msg))
+                    except:
+                        print("Wrong formatted msg.")
```

Comment:
Looks better. `json.JSONDecodeError` only exists in >= 3.5, below it's ValueError. As JSONDecodeError is a subclass of ValueError, catching only the latter one is fine.

The default type would be either Report or Event, based on the bot's group (the parameter "group"). Then the user does not need to give the message type at all and it's always correct :)

Also, for bot's without destination pipeline this throws an exception:
```
...
file-output: Opening '/opt/intelmq/var/lib/bots/file-output/events.txt' file.
file-output: File '/opt/intelmq/var/lib/bots/file-output/events.txt' is open.
file-output: Loading source pipeline and queue 'file-output-queue'.
file-output: Connected to source queue.
file-output: No destination queues to load.
file-output: Pipeline ready.
Traceback (most recent call last):
  File "/usr/local/bin/intelmqctl", line 9, in <module>
    load_entry_point('intelmq==1.0.0.dev7', 'console_scripts', 'intelmqctl')()
  File "/home/sebastian/dev/intelmq/intelmq/bin/intelmqctl.py", line 885, in main
    return x.run()
  File "/home/sebastian/dev/intelmq/intelmq/bin/intelmqctl.py", line 531, in run
    results = args.func(**args_dict)
  File "/home/sebastian/dev/intelmq/intelmq/bin/intelmqctl.py", line 541, in bot_run
    return self.bot_process_manager.bot_run(bot_id, run_subcommand, message_action_kind, dryrun, msg)
  File "/home/sebastian/dev/intelmq/intelmq/bin/intelmqctl.py", line 135, in bot_run
    BotDebugger(self.__runtime_configuration[bot_id]['module'], bot_id, run_subcommand, message_action_kind, dryrun, msg)
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot_debugger.py", line 45, in __init__
    self._message(message_kind, msg)
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot_debugger.py", line 80, in _message
    self.instance.send_message(msg)
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 332, in send_message
    raise exceptions.ConfigurationError('pipeline', 'No destination pipeline given, '
intelmq.lib.exceptions.ConfigurationError: pipeline configuration failed - No destination pipeline given, but needed
```